Fix the app:build command on windows

shadoWalker89 opened the pull request:
This PR fixes a problem when building the .phar on windows.

The error is ` ErrorException  : rename(W:\projects\app-name\app-name.phar,W:\projects\app-name\builds\app-name): Le fichier spcifi est introuvable. (code: 2)`

This due to the fact that compiling the .phar didn't happen because `./box` doesn't work on windows.

I also tested building the app on a ubuntu machine after removing the `./` and it works

shadoWalker89 commented:
The failing tests are not caused by the PR but by some composer dependencies

nunomaduro commented:
@ijpatricio Can you check this one?

ijpatricio commented:
Hey @nunomaduro sute, glad I can help.

@shadoWalker89 was this your only change, for sure?

Look at this picture below. I applied the one change in this PR, and getting the same result with Powershell and Git Bash

![image](https://user-images.githubusercontent.com/26031459/55195072-57703000-51a3-11e9-9c12-94a4f1f62c69.png)

ijpatricio commented:
Also, I dowloaded lastest version from https://github.com/humbug/box, whici is 3.6 (laravel-zero is now with 3.5.1) and result is the same

[humbug/box is having issues on windows, check here](https://github.com/humbug/box/issues/193)

`Filename syntax, directory name, or volume label is incorrect`

![image](https://user-images.githubusercontent.com/26031459/55196453-55a86b80-51a7-11e9-9213-9cb5b5ced154.png)

shadoWalker89 commented:
@ijpatricio Yeah, that is the change that worked for me. But now  i know why it didn't work for you.

Turns out running `box` on windows will not work.

It worked for me because i installed the box package as a global composer dependency.

So i checked the `humbug/box` and i saw that they are using a `box.bat` file that calls the `box` file

So i removed the `humbug/box` from my composer global dependencies to avoid having false positives.
I added a .bat file with the following content at `vendor\laravel-zero\framework\bin`

```
@ECHO OFF
php %~dp0box %*
```
And also update the build command with this line

```php
(! windows_os() ? './box' : 'box.bat').' compile --working-dir="'.base_path().'" --config="'.base_path('box.json').'"',
```

The app:build command still fails, but this time it's not because of the box command not getting executed but because of something else. If you try and run the app:build with -vvv you will probably get the same error that i got

`La syntaxe du nom de fichier, de répertoire ou de volume est incorrecte.`
English version of it
`The filename, directory name, or volume label syntax is incorrect`

I think this probably has something to do with the `box` file included with laravel-zero not compatible with windows.

So what works, and i tried, and i hope that you can confirm that is installing the `humbug/box` as a composer dependency and call the .bat file included with it.

I updated the PR with the code needed for it to work.
For windows we call the .bat file from the vendor `/bin` directory and for the other platforms we use the original behavior.

For this to work of course you must install `humbug/box` as a composer dependency on the project.

If this PR is accepted, i will make a PR to the application laravel-zero/laravel-zero repository to add the `humbug/box` as a dev dependency

c33s commented on src/Commands/BuildCommand.php:
```diff
@@ -100,8 +100,8 @@ private function compile(string $name): BuildCommand
         $process = new Process(
-            './box compile --working-dir="'.base_path().'" --config="'.base_path('box.json').'"',
-            dirname(dirname(__DIR__)).'/bin',
+            (! windows_os() ? './box' : 'box.bat').' compile --working-dir="'.base_path().'" --config="'.base_path('box.json').'"',
```
just a suggestion, as i find the code quite unreadable with the ternary operator, i would prefer to have the code structured like that:

```php
<?php
if (! windows_os()) {
    $boxCommand = './box';
    $workDir = dirname(dirname(__DIR__)).'/bin'
} else {
    $boxCommand = 'box.bat';
    $workDir = base_path('vendor')
}
 
$process = new Process($boxCommand.' compile --working-dir="'.base_path().'" --config="'.base_path('box.json').'"',           $workDir, ...
```
haven't tested the this PR but `base_path('vendor')` look a little strange to me (maybe i don't understand the problem).  does this produce a correct path if called like that `php ./myproject/box/bin/console app:build`?

nunomaduro commented:
@shadoWalker89 Is this pull request ready for merge?

pxgamer commented:
Just been looking into this while updating Box to v3.8.3, I created a similar commit cccc0e0d803bbaa7d38094b05ed2730ecf4b080d which adds a local `box.bat` file and doesn't require the directory switch. Would this be a good idea to add? 🤔

> I didn't know about the `windows_os()` helper, so it can easily be changed to:
> ```php
> $boxBinary = windows_os() ? 'box' : './box';
> ```

pxgamer commented:
Hi @shadoWalker89, I feel that using the local Box binary would be better for consistency across Laravel Zero. Would you be able to create a local `box.bat`, similar to what I did in cccc0e0d803bbaa7d38094b05ed2730ecf4b080d?

If you'd like me to make the changes, I'd be happy to do this. 👍

theofidry commented:
If that makes sense the bat fine can be changed in Box itself. There is actually one there but I never used it so I don't know if it works (as I don't have a Windows machine either)

pxgamer commented:
Sorry for the delay, I feel like the `box.bat` should be added here, as Laravel Zero directly includes the Phar. I've just re-tested a combination of my `.bat`, and the changes here, and I'm now getting the `File or directory name is invalid` error again. Looks like this can be resolved by setting ` --no-restart` in the command (i.e. `box.bat --no-restart` for the `$boxBinary`), however it would be nice to get Windows supported in Box (as per https://github.com/humbug/box/issues/193) without needing this.

owenvoke commented:
Hi @shadoWalker89, thanks for the work you've done on this. These changes have been merged into v7.x of the framework so I'm going to close this. Cheers.

If you experience any problems, please open an issue. 👍
